File: native-ios/Solvio/Core/UI/VendorLogo.swift
import SwiftUI

/// Brand-logo badge for a vendor (Lidl, Biedronka, Kaufland…). Looks
/// up the vendor name in a curated map of Polish + global retailers,
/// fetches the favicon via DuckDuckGo's icon service, and renders
/// inside the same neobrutalist tile shape as `NBIconBadge`. Falls
/// back to `NBIconBadge(systemImage:)` when the vendor doesn't match
/// any known chain or the network image hasn't loaded yet.
///
/// Why DuckDuckGo: their `icons.duckduckgo.com/ip3/<domain>.ico`
/// endpoint is free, no API key, returns reliable favicons for any
/// public domain, and crucially returns a transparent PNG variant for
/// a lot of brands (so the logo blends with the muted card backdrop
/// instead of carrying a coloured square).
struct VendorLogo: View {
    let vendor: String?
    var size: CGFloat = 36
    /// SF Symbol name to fall back to while loading or when vendor
    /// can't be matched. Defaults to a generic shopping-bag glyph.
    var fallbackIcon: String = "bag.fill"

    var body: some View {
        if let domain = Self.domain(for: vendor),
           let url = URL(string: "https://icons.duckduckgo.com/ip3/\(domain).ico") {
            AsyncImage(url: url, transaction: Transaction(animation: .easeOut(duration: 0.2))) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .scaledToFit()
                        .padding(size * 0.16)
                        .frame(width: size, height: size)
                        .background(Theme.muted)
                        .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.sm))
                        .overlay(
                            RoundedRectangle(cornerRadius: Theme.Radius.sm)
                                .stroke(Theme.border, lineWidth: Theme.Border.widthThin)
                        )
                case .failure, .empty:
                    NBIconBadge(systemImage: fallbackIcon, size: size)
                @unknown default:
                    NBIconBadge(systemImage: fallbackIcon, size: size)
                }
            }
        } else {
            NBIconBadge(systemImage: fallbackIcon, size: size)
        }
    }

    /// Returns the canonical web domain for a vendor name, or `nil`
    /// when no match. Match is case-insensitive + handles common
    /// suffix variants ("Lidl Sp. z o.o.", "Biedronka 1234", etc.).
    static func domain(for vendor: String?) -> String? {
        guard let v = vendor?.lowercased().trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty else {
            return nil
        }
        // Exact match first (cheaper than scanning the contains map).
        if let direct = exactMap[v] { return direct }
        // Then look for any chain key that appears as a token / substring.
        for (key, dom) in containsMap {
            if v.contains(key) { return dom }
        }
        return nil
    }

    // MARK: - Lookup tables

    /// Vendor strings that match exactly (after lowercasing). Add
    /// here when a brand name is short enough to safely identity-match
    /// without ambiguity.
    private static let exactMap: [String: String] = [
        "lidl": "lidl.pl",
        "biedronka": "biedronka.pl",
        "kaufland": "kaufland.pl",
        "auchan": "auchan.pl",
        "carrefour": "carrefour.pl",
        "netto": "netto.pl",
        "dino": "grupadino.pl",
        "stokrotka": "stokrotka.pl",
        "aldi": "aldi.pl",
        "żabka": "zabka.pl",
        "zabka": "zabka.pl",
        "rossmann": "rossmann.pl",
        "hebe": "hebe.pl",
        "ikea": "ikea.com",
        "decathlon": "decathlon.pl",
        "leroy merlin": "leroymerlin.pl",
        "obi": "obi.pl",
        "castorama": "castorama.pl",
        "polomarket": "polomarket.pl",
        "delikatesy centrum": "delikatesy.pl",
        "intermarche": "intermarche.pl",
        "tesco": "tesco.pl",
        "spar": "spar.pl",
        "freshmarket": "freshmarket.pl",
        "primark": "primark.com",
        "h&m": "hm.com",
        "zara": "zara.com",
        "reserved": "reserved.com",
        "cropp": "cropp.com",
        "mohito": "mohito.com",
        "house": "housebrand.com",
        "sinsay": "sinsay.com",
        "mediaexpert": "mediaexpert.pl",
        "media markt": "mediamarkt.pl",
        "mediamarkt": "mediamarkt.pl",
        "rtv euro agd": "euro.com.pl",
        "x-kom": "x-kom.pl",
        "morele": "morele.net",
        "allegro": "allegro.pl",
        "amazon": "amazon.pl",
        "empik": "empik.com",
        "smyk": "smyk.com",
        "rebel": "rebel.pl",
        "pepco": "pepco.pl",
        "kik": "kik.pl",
        "tedi": "tedi.com",
        "action": "action.com",
        "drogerie natura": "drogerie-natura.pl",
        "super-pharm": "super-pharm.pl",
        "uber": "uber.com",
        "ubereats": "ubereats.com",
        "uber eats": "ubereats.com",
        "wolt": "wolt.com",
        "bolt": "bolt.eu",
        "free now": "free-now.com",
        "spotify": "spotify.com",
        "netflix": "netflix.com",
        "youtube": "youtube.com",
        "youtube premium": "youtube.com",
        "apple": "apple.com",
        "icloud": "apple.com",
        "google": "google.com",
        "google one": "google.com",
        "microsoft": "microsoft.com",
        "office 365": "microsoft.com",
        "orange": "orange.pl",
        "play": "play.pl",
        "t-mobile": "t-mobile.pl",
        "plus": "plus.pl",
        "pgnig": "pgnig.pl",
        "tauron": "tauron.pl",
        "innogy": "innogy.pl",
        "veolia": "veolia.com",
        "pkn orlen": "orlen.pl",
        "orlen": "orlen.pl",
        "bp": "bp.com",
        "shell": "shell.com",
        "circle k": "circlek.com",
        "moya": "moya.pl",
        "jysk": "jysk.com",
        "agata": "agatameble.pl",
        "black red white": "br-w.pl",
        "vox": "vox.pl",
        "starbucks": "starbucks.pl",
        "mcdonald's": "mcdonalds.pl",
        "mcdonalds": "mcdonalds.pl",
        "kfc": "kfc.pl",
        "burger king": "burgerking.pl",
        "subway": "subway.com",
        "pizza hut": "pizzahut.pl",
        "domino's": "dominos.pl",
        "dominos": "dominos.pl",
        "telepizza": "telepizza.pl",
    ]

    /// Substring matches. `vendor.contains(key)` returns the domain.
    /// Order matters: more specific keys first (so "lidl plus" hits
    /// before bare "lidl"). Keep concise — the more entries, the more
    /// false positives.
    private static let containsMap: [(String, String)] = [
        ("lidl", "lidl.pl"),
        ("biedronka", "biedronka.pl"),
        ("kaufland", "kaufland.pl"),
        ("auchan", "auchan.pl"),
        ("carrefour", "carrefour.pl"),
        ("netto", "netto.pl"),
        ("dino", "grupadino.pl"),
        ("stokrotka", "stokrotka.pl"),
        ("aldi", "aldi.pl"),
        ("żabka", "zabka.pl"),
        ("zabka", "zabka.pl"),
        ("rossmann", "rossmann.pl"),
        ("hebe", "hebe.pl"),
        ("ikea", "ikea.com"),
        ("decathlon", "decathlon.pl"),
        ("leroy merlin", "leroymerlin.pl"),
        ("castorama", "castorama.pl"),
        ("polomarket", "polomarket.pl"),
        ("polo market", "polomarket.pl"),
        ("media markt", "mediamarkt.pl"),
        ("mediamarkt", "mediamarkt.pl"),
        ("media expert", "mediaexpert.pl"),
        ("mediaexpert", "mediaexpert.pl"),
        ("rtv euro agd", "euro.com.pl"),
        ("euro agd", "euro.com.pl"),
        ("x-kom", "x-kom.pl"),
        ("xkom", "x-kom.pl"),
        ("allegro", "allegro.pl"),
        ("amazon", "amazon.pl"),
        ("empik", "empik.com"),
        ("smyk", "smyk.com"),
        ("pepco", "pepco.pl"),
        ("super-pharm", "super-pharm.pl"),
        ("super pharm", "super-pharm.pl"),
        ("uber eats", "ubereats.com"),
        ("ubereats", "ubereats.com"),
        ("uber", "uber.com"),
        ("wolt", "wolt.com"),
        ("bolt", "bolt.eu"),
        ("spotify", "spotify.com"),
        ("netflix", "netflix.com"),
        ("youtube", "youtube.com"),
        ("apple", "apple.com"),
        ("icloud", "apple.com"),
        ("google", "google.com"),
        ("microsoft", "microsoft.com"),
        ("office", "microsoft.com"),
        ("orange", "orange.pl"),
        ("t-mobile", "t-mobile.pl"),
        ("orlen", "orlen.pl"),
        ("shell", "shell.com"),
        ("bp polska", "bp.com"),
        ("circle k", "circlek.com"),
        ("jysk", "jysk.com"),
        ("starbucks", "starbucks.pl"),
        ("mcdonald", "mcdonalds.pl"),
        ("kfc", "kfc.pl"),
        ("burger king", "burgerking.pl"),
        ("pizza hut", "pizzahut.pl"),
        ("domino", "dominos.pl"),
        ("telepizza", "telepizza.pl"),
        ("h&m", "hm.com"),
        ("h & m", "hm.com"),
        ("zara", "zara.com"),
        ("reserved", "reserved.com"),
        ("cropp", "cropp.com"),
        ("sinsay", "sinsay.com"),
        ("primark", "primark.com"),
        ("rebel", "rebel.pl"),
        ("intermarche", "intermarche.pl"),
        ("intermarché", "intermarche.pl"),
        ("tesco", "tesco.pl"),
    ]
}

File: android/app/src/main/java/com/programo/solvio/core/ui/VendorLogo.kt
package com.programo.solvio.core.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ShoppingBag
import androidx.compose.ui.graphics.vector.ImageVector
import coil.compose.AsyncImagePainter
import coil.compose.SubcomposeAsyncImage
import coil.compose.SubcomposeAsyncImageContent
import coil.request.ImageRequest
import androidx.compose.ui.platform.LocalContext
import com.programo.solvio.core.theme.LocalPalette
import com.programo.solvio.core.theme.SolvioTheme

/// Brand-logo badge for a vendor (Lidl, Biedronka, Kaufland…). Faithful
/// Compose port of `native-ios/Solvio/Core/UI/VendorLogo.swift`. Looks
/// up the vendor name in a curated map of Polish + global retailers,
/// fetches the favicon via DuckDuckGo's icon service, and renders inside
/// the same neobrutalist tile shape as [NBIconBadge]. Falls back to an
/// [NBIconBadge] symbol when the vendor can't be matched or the network
/// image hasn't loaded / fails.
///
/// Why DuckDuckGo: `icons.duckduckgo.com/ip3/<domain>.ico` is free, no
/// API key, and returns reliable (often transparent) favicons that blend
/// with the muted card backdrop.
@Composable
fun VendorLogo(
    vendor: String?,
    size: Dp = 36.dp,
    fallbackIcon: ImageVector = Icons.Filled.ShoppingBag,
) {
    val palette = LocalPalette.current
    val domain = VendorLogoMap.domain(vendor)

    if (domain == null) {
        NBIconBadge(icon = fallbackIcon, size = size)
        return
    }

    val context = LocalContext.current
    val request = ImageRequest.Builder(context)
        .data("https://icons.duckduckgo.com/ip3/$domain.ico")
        .crossfade(true)
        .build()

    SubcomposeAsyncImage(
        model = request,
        contentDescription = vendor,
    ) {
        when (painter.state) {
            is AsyncImagePainter.State.Success -> {
                Box(
                    modifier = Modifier
                        .size(size)
                        .clip(RoundedCornerShape(SolvioTheme.Radius.sm))
                        .background(palette.muted)
                        .border(
                            SolvioTheme.Border.widthThin,
                            palette.border,
                            RoundedCornerShape(SolvioTheme.Radius.sm),
                        ),
                    contentAlignment = Alignment.Center,
                ) {
                    SubcomposeAsyncImageContent(
                        modifier = Modifier
                            .size(size)
                            .padding(size * 0.16f),
                        contentScale = ContentScale.Fit,
                    )
                }
            }
            else -> NBIconBadge(icon = fallbackIcon, size = size)
        }
    }
}

/// Vendor-name → web-domain resolver. 1:1 port of `VendorLogo.exactMap`
/// + `VendorLogo.containsMap` from the iOS source. Match is
/// case-insensitive and handles common suffix variants
/// ("Lidl Sp. z o.o.", "Biedronka 1234", …).
object VendorLogoMap {
    fun domain(vendor: String?): String? {
        val v = vendor?.lowercase()?.trim()
        if (v.isNullOrEmpty()) return null
        exactMap[v]?.let { return it }
        for ((key, dom) in containsMap) {
            if (v.contains(key)) return dom
        }
        return null
    }

    private val exactMap: Map<String, String> = mapOf(
        "lidl" to "lidl.pl",
        "biedronka" to "biedronka.pl",
        "kaufland" to "kaufland.pl",
        "auchan" to "auchan.pl",
        "carrefour" to "carrefour.pl",
        "netto" to "netto.pl",
        "dino" to "grupadino.pl",
        "stokrotka" to "stokrotka.pl",
        "aldi" to "aldi.pl",
        "żabka" to "zabka.pl",
        "zabka" to "zabka.pl",
        "rossmann" to "rossmann.pl",
        "hebe" to "hebe.pl",
        "ikea" to "ikea.com",
        "decathlon" to "decathlon.pl",
        "leroy merlin" to "leroymerlin.pl",
        "obi" to "obi.pl",
        "castorama" to "castorama.pl",
        "polomarket" to "polomarket.pl",
        "delikatesy centrum" to "delikatesy.pl",
        "intermarche" to "intermarche.pl",
        "tesco" to "tesco.pl",
        "spar" to "spar.pl",
        "freshmarket" to "freshmarket.pl",
        "primark" to "primark.com",
        "h&m" to "hm.com",
        "zara" to "zara.com",
        "reserved" to "reserved.com",
        "cropp" to "cropp.com",
        "mohito" to "mohito.com",
        "house" to "housebrand.com",
        "sinsay" to "sinsay.com",
        "mediaexpert" to "mediaexpert.pl",
        "media markt" to "mediamarkt.pl",
        "mediamarkt" to "mediamarkt.pl",
        "rtv euro agd" to "euro.com.pl",
        "x-kom" to "x-kom.pl",
        "morele" to "morele.net",
        "allegro" to "allegro.pl",
        "amazon" to "amazon.pl",
        "empik" to "empik.com",
        "smyk" to "smyk.com",
        "rebel" to "rebel.pl",
        "pepco" to "pepco.pl",
        "kik" to "kik.pl",
        "tedi" to "tedi.com",
        "action" to "action.com",
        "drogerie natura" to "drogerie-natura.pl",
        "super-pharm" to "super-pharm.pl",
        "uber" to "uber.com",
        "ubereats" to "ubereats.com",
        "uber eats" to "ubereats.com",
        "wolt" to "wolt.com",
        "bolt" to "bolt.eu",
        "free now" to "free-now.com",
        "spotify" to "spotify.com",
        "netflix" to "netflix.com",
        "youtube" to "youtube.com",
        "youtube premium" to "youtube.com",
        "apple" to "apple.com",
        "icloud" to "apple.com",
        "google" to "google.com",
        "google one" to "google.com",
        "microsoft" to "microsoft.com",
        "office 365" to "microsoft.com",
        "orange" to "orange.pl",
        "play" to "play.pl",
        "t-mobile" to "t-mobile.pl",
        "plus" to "plus.pl",
        "pgnig" to "pgnig.pl",
        "tauron" to "tauron.pl",
        "innogy" to "innogy.pl",
        "veolia" to "veolia.com",
        "pkn orlen" to "orlen.pl",
        "orlen" to "orlen.pl",
        "bp" to "bp.com",
        "shell" to "shell.com",
        "circle k" to "circlek.com",
        "moya" to "moya.pl",
        "jysk" to "jysk.com",
        "agata" to "agatameble.pl",
        "black red white" to "br-w.pl",
        "vox" to "vox.pl",
        "starbucks" to "starbucks.pl",
        "mcdonald's" to "mcdonalds.pl",
        "mcdonalds" to "mcdonalds.pl",
        "kfc" to "kfc.pl",
        "burger king" to "burgerking.pl",
        "subway" to "subway.com",
        "pizza hut" to "pizzahut.pl",
        "domino's" to "dominos.pl",
        "dominos" to "dominos.pl",
        "telepizza" to "telepizza.pl",
    )

    /// Substring matches — order matters (more specific keys first).
    private val containsMap: List<Pair<String, String>> = listOf(
        "lidl" to "lidl.pl",
        "biedronka" to "biedronka.pl",
        "kaufland" to "kaufland.pl",
        "auchan" to "auchan.pl",
        "carrefour" to "carrefour.pl",
        "netto" to "netto.pl",
        "dino" to "grupadino.pl",
        "stokrotka" to "stokrotka.pl",
        "aldi" to "aldi.pl",
        "żabka" to "zabka.pl",
        "zabka" to "zabka.pl",
        "rossmann" to "rossmann.pl",
        "hebe" to "hebe.pl",
        "ikea" to "ikea.com",
        "decathlon" to "decathlon.pl",
        "leroy merlin" to "leroymerlin.pl",
        "castorama" to "castorama.pl",
        "polomarket" to "polomarket.pl",
        "polo market" to "polomarket.pl",
        "media markt" to "mediamarkt.pl",
        "mediamarkt" to "mediamarkt.pl",
        "media expert" to "mediaexpert.pl",
        "mediaexpert" to "mediaexpert.pl",
        "rtv euro agd" to "euro.com.pl",
        "euro agd" to "euro.com.pl",
        "x-kom" to "x-kom.pl",
        "xkom" to "x-kom.pl",
        "allegro" to "allegro.pl",
        "amazon" to "amazon.pl",
        "empik" to "empik.com",
        "smyk" to "smyk.com",
        "pepco" to "pepco.pl",
        "super-pharm" to "super-pharm.pl",
        "super pharm" to "super-pharm.pl",
        "uber eats" to "ubereats.com",
        "ubereats" to "ubereats.com",
        "uber" to "uber.com",
        "wolt" to "wolt.com",
        "bolt" to "bolt.eu",
        "spotify" to "spotify.com",
        "netflix" to "netflix.com",
        "youtube" to "youtube.com",
        "apple" to "apple.com",
        "icloud" to "apple.com",
        "google" to "google.com",
        "microsoft" to "microsoft.com",
        "office" to "microsoft.com",
        "orange" to "orange.pl",
        "t-mobile" to "t-mobile.pl",
        "orlen" to "orlen.pl",
        "shell" to "shell.com",
        "bp polska" to "bp.com",
        "circle k" to "circlek.com",
        "jysk" to "jysk.com",
        "starbucks" to "starbucks.pl",
        "mcdonald" to "mcdonalds.pl",
        "kfc" to "kfc.pl",
        "burger king" to "burgerking.pl",
        "pizza hut" to "pizzahut.pl",
        "domino" to "dominos.pl",
        "telepizza" to "telepizza.pl",
        "h&m" to "hm.com",
        "h & m" to "hm.com",
        "zara" to "zara.com",
        "reserved" to "reserved.com",
        "cropp" to "cropp.com",
        "sinsay" to "sinsay.com",
        "primark" to "primark.com",
        "rebel" to "rebel.pl",
        "intermarche" to "intermarche.pl",
        "intermarché" to "intermarche.pl",
        "tesco" to "tesco.pl",
    )
}
